route-snapper/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+                    .to_polygon()
+                    .to_geojson(gps_bounds),
+                make_props("type", "confirmed route intersection"),
+            ));
+        }
+
+        // Draw the current operation
+        if let Mode::Hovering(i) = self.mode {
+            pairs.push((
+                Circle::new(self.map.get_i(i).polygon.center(), INTERSECTON_RADIUS)
+                    .to_polygon()
+                    .to_geojson(gps_bounds),
+                make_props("type", "hovering intersection"),
+            ));
+            if self.route.waypoints.len() == 1 {
+                if let Some((roads, intersections)) =
+                    self.map
+                        .simple_path_btwn_v2(self.route.waypoints[0], i, PathConstraints::Car)
+                {
+                    for r in roads {
+                        let r = self.map.get_r(r);
+                        pairs.push((
+                            r.center_pts.to_geojson(gps_bounds),
+                            make_props("type", "preview route leg"),
+                        ));
+                    }
+                    for i in intersections {
+                        pairs.push((
+                            Circle::new(self.map.get_i(i).polygon.center(), INTERSECTON_RADIUS)
+                                .to_polygon()
+                                .to_geojson(gps_bounds),
+                            make_props("type", "preview intersection"),
+                        ));
+                    }
+                }
+            }
+        }
+        if let Mode::Dragging { at, .. } = self.mode {
+            pairs.push((
+                Circle::new(self.map.get_i(at).polygon.center(), INTERSECTON_RADIUS)
+                    .to_polygon()
+                    .to_geojson(gps_bounds),
+                make_props("type", "drag intersection"),
+            ));
+        }
+
+        let obj = geom::geometries_with_properties_to_geojson(pairs);
+        serde_json::to_string_pretty(&obj).unwrap()
+    }
+
+    #[wasm_bindgen(js_name = toFinalFeature)]
+    pub fn to_final_feature(&self) -> Option<String> {
+        let mut pts = Vec::new();
+        for pair in self.route.full_path.windows(2) {
+            let r = self
+                .map
+                .get_r(self.map.find_road_between(pair[0], pair[1]).unwrap());
+            pts.extend(r.center_pts.clone().into_points());
+        }
+        let pl = PolyLine::deduping_new(pts).ok()?;
+        let feature = geojson::Feature {
+            bbox: None,
+            geometry: Some(pl.to_geojson(Some(self.map.get_gps_bounds()))),
+            id: Some(geojson::feature::Id::String("snappy-route".to_string())),
+            properties: None,
+            foreign_members: None,
+        };
+        Some(serde_json::to_string_pretty(&feature).unwrap())
+    }
+
+    // True if something has changed
+    #[wasm_bindgen(js_name = onMouseMove)]
+    pub fn on_mouse_move(&mut self, lon: f64, lat: f64) -> bool {
```

Comment:
There's probably too much UI logic in here. mapbox draw plugins handle this kind of thing in more natural ways. The responsibility of the Rust lib should probably shrink dramatically